File: java/shared/java/org/signal/libsignal/zkgroup/profiles/ClientZkProfileOperations.java
//
// Copyright 2020-2022 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.profiles;

import static org.signal.libsignal.internal.FilterExceptions.filterExceptions;
import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

import java.security.SecureRandom;
import java.time.Instant;
import org.signal.libsignal.internal.Native;
import org.signal.libsignal.protocol.ServiceId.Aci;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.ServerPublicParams;
import org.signal.libsignal.zkgroup.VerificationFailedException;
import org.signal.libsignal.zkgroup.groups.GroupSecretParams;

public class ClientZkProfileOperations {

  private final ServerPublicParams serverPublicParams;

  public ClientZkProfileOperations(ServerPublicParams serverPublicParams) {
    this.serverPublicParams = serverPublicParams;
  }

  public ProfileKeyCredentialRequestContext createProfileKeyCredentialRequestContext(
      Aci userId, ProfileKey profileKey) {
    return createProfileKeyCredentialRequestContext(new SecureRandom(), userId, profileKey);
  }

  public ProfileKeyCredentialRequestContext createProfileKeyCredentialRequestContext(
      SecureRandom secureRandom, Aci userId, ProfileKey profileKey) {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents =
        serverPublicParams.guardedMap(
            (serverPublicParams) ->
                Native.ServerPublicParams_CreateProfileKeyCredentialRequestContextDeterministic(
                    serverPublicParams,
                    random,
                    userId.toServiceIdFixedWidthBinary(),
                    profileKey.getInternalContentsForJNI()));

    try {
      return new ProfileKeyCredentialRequestContext(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ExpiringProfileKeyCredential receiveExpiringProfileKeyCredential(
      ProfileKeyCredentialRequestContext profileKeyCredentialRequestContext,
      ExpiringProfileKeyCredentialResponse profileKeyCredentialResponse)
      throws VerificationFailedException {
    return receiveExpiringProfileKeyCredential(
        profileKeyCredentialRequestContext, profileKeyCredentialResponse, Instant.now());
  }

  public ExpiringProfileKeyCredential receiveExpiringProfileKeyCredential(
      ProfileKeyCredentialRequestContext profileKeyCredentialRequestContext,
      ExpiringProfileKeyCredentialResponse profileKeyCredentialResponse,
      Instant now)
      throws VerificationFailedException {
    if (profileKeyCredentialResponse == null) {
      throw new VerificationFailedException();
    }

    byte[] newContents =
        filterExceptions(
            VerificationFailedException.class,
            () ->
                serverPublicParams.guardedMapChecked(
                    (publicParams) ->
                        Native.ServerPublicParams_ReceiveExpiringProfileKeyCredential(
                            publicParams,
                            profileKeyCredentialRequestContext.getInternalContentsForJNI(),
                            profileKeyCredentialResponse.getInternalContentsForJNI(),
                            now.getEpochSecond())));

    try {
      return new ExpiringProfileKeyCredential(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public ProfileKeyCredentialPresentation createProfileKeyCredentialPresentation(
      GroupSecretParams groupSecretParams, ExpiringProfileKeyCredential profileKeyCredential) {
    return createProfileKeyCredentialPresentation(
        new SecureRandom(), groupSecretParams, profileKeyCredential);
  }

  public ProfileKeyCredentialPresentation createProfileKeyCredentialPresentation(
      SecureRandom secureRandom,
      GroupSecretParams groupSecretParams,
      ExpiringProfileKeyCredential profileKeyCredential) {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents =
        serverPublicParams.guardedMap(
            (publicParams) ->
                Native
                    .ServerPublicParams_CreateExpiringProfileKeyCredentialPresentationDeterministic(
                        publicParams,
                        random,
                        groupSecretParams.getInternalContentsForJNI(),
                        profileKeyCredential.getInternalContentsForJNI()));

    try {
      return new ProfileKeyCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}

File: swift/Sources/LibSignalClient/zkgroup/ClientZkProfileOperations.swift
//
// Copyright 2020-2022 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public class ClientZkProfileOperations {
    let serverPublicParams: ServerPublicParams

    public init(serverPublicParams: ServerPublicParams) {
        self.serverPublicParams = serverPublicParams
    }

    public func createProfileKeyCredentialRequestContext(userId: Aci, profileKey: ProfileKey) throws -> ProfileKeyCredentialRequestContext {
        return try self.createProfileKeyCredentialRequestContext(randomness: Randomness.generate(), userId: userId, profileKey: profileKey)
    }

    public func createProfileKeyCredentialRequestContext(randomness: Randomness, userId: Aci, profileKey: ProfileKey) throws -> ProfileKeyCredentialRequestContext {
        return try self.serverPublicParams.withNativeHandle { serverPublicParams in
            try randomness.withUnsafePointerToBytes { randomness in
                try userId.withPointerToFixedWidthBinary { userId in
                    try profileKey.withUnsafePointerToSerialized { profileKey in
                        try invokeFnReturningSerialized {
                            signal_server_public_params_create_profile_key_credential_request_context_deterministic($0, serverPublicParams, randomness, userId, profileKey)
                        }
                    }
                }
            }
        }
    }

    public func receiveExpiringProfileKeyCredential(
        profileKeyCredentialRequestContext: ProfileKeyCredentialRequestContext,
        profileKeyCredentialResponse: ExpiringProfileKeyCredentialResponse,
        now: Date = Date()
    ) throws -> ExpiringProfileKeyCredential {
        return try self.serverPublicParams.withNativeHandle { serverPublicParams in
            try profileKeyCredentialRequestContext.withUnsafePointerToSerialized { requestContext in
                try profileKeyCredentialResponse.withUnsafePointerToSerialized { response in
                    try invokeFnReturningSerialized {
                        signal_server_public_params_receive_expiring_profile_key_credential($0, serverPublicParams, requestContext, response, UInt64(now.timeIntervalSince1970))
                    }
                }
            }
        }
    }

    public func createProfileKeyCredentialPresentation(groupSecretParams: GroupSecretParams, profileKeyCredential: ExpiringProfileKeyCredential) throws -> ProfileKeyCredentialPresentation {
        return try self.createProfileKeyCredentialPresentation(randomness: Randomness.generate(), groupSecretParams: groupSecretParams, profileKeyCredential: profileKeyCredential)
    }

    public func createProfileKeyCredentialPresentation(randomness: Randomness, groupSecretParams: GroupSecretParams, profileKeyCredential: ExpiringProfileKeyCredential) throws -> ProfileKeyCredentialPresentation {
        return try self.serverPublicParams.withNativeHandle { serverPublicParams in
            try randomness.withUnsafePointerToBytes { randomness in
                try groupSecretParams.withUnsafePointerToSerialized { groupSecretParams in
                    try profileKeyCredential.withUnsafePointerToSerialized { profileKeyCredential in
                        try invokeFnReturningVariableLengthSerialized {
                            signal_server_public_params_create_expiring_profile_key_credential_presentation_deterministic($0, serverPublicParams, randomness, groupSecretParams, profileKeyCredential)
                        }
                    }
                }
            }
        }
    }
}
